runtime/runtime/src/actions.rs

```diff
@@ -613,6 +615,46 @@ pub(crate) fn action_add_key(
     Ok(())
 }
 
+pub(crate) fn action_delegate_action(
+    apply_state: &ApplyState,
+    action_receipt: &ActionReceipt,
+    predecessor_id: &AccountId,
+    signed_delegate_action: &SignedDelegateAction,
+    result: &mut ActionResult,
+) -> Result<(), RuntimeError> {
+    match signed_delegate_action.get_delegate_action() {
+        Ok(delegate_action) => {
+            let new_receipt = Receipt::new_delegate_actions(
+                &action_receipt.signer_id,
+                predecessor_id,
+                &delegate_action,
+                &signed_delegate_action.public_key,
+                action_receipt.gas_price,
+            );
+
+            let transaction_costs = &apply_state.config.transaction_costs;
+            let current_protocol_version = apply_state.current_protocol_version;
+            let cost = receipt_cost(transaction_costs, current_protocol_version, &new_receipt)?;
+
+            if let Some(refund) = delegate_action.deposit.checked_sub(cost.clone()) {
+                let refund_receipt = Receipt::new_balance_refund(&action_receipt.signer_id, refund);
```

Comment:
> Is there a reason that we have to handle deposit refunds specially? I would have expected that we treat it like existing deposit refunds. That is, towards the end of `apply_action_receipt` we call `generate_refund_receipts` and give a refund to the predecessor of the action. The predecessor of the delegation action should be the publisher, right? So that seems correct to me.

This deposit covers the total cost (gas and deposits) of all actions in DelegateAction. To run actions from DelegateAction, some prepaid gas is required. Relayer doesn't prepaid gas but attaches deposit instead. If Relayer attaches deposit which is less than the total cost of actions, need to generate a refund.
`generate_refund_receipts` generates a refund only when the receipt fails. This is not a fail case.

Why just do not pre-pay gas? Then I will play with `result.gas_used` counter here (and maybe deposit, if we want to refund it).  `result.gas_used` will not be actually used but will be "reserved" for actions in DelegateAction. Otherwise `generate_refund_receipts` will refund all prepaid gas and `check_balance` call panic.

Both variants are tricky and I chose the first one.

Reply:
I see, thanks for explaining.

Then, there is one issue with the current design. The gas price for the initial receipts is computed based on attached gas = 0 because the delegate action has no attached gas. This gas price is then passed on to the inner actions receipt produced by the relayer. This circumvents the pessimistic price calculation that is meant to ensure that no receipt ever executed at a gas price lower than the current gas price, even if the gas price is increased on every block.

Prepaying all the gas upfront would avoid such trickiness. And it would be more in line with the existing bahaviour.

The protocol currently has two situations where gas is burnt later than when it is purchased.

1. In general, action receipts are created in one block and executed in the next. In `ActionResult`,  `gas_burnt` covers the cost for executing the current action and for creating new receipts without executing them. `gas_used` covers the full cost of executing the current action and all new action receipts. In other words, `gas_used` is `gas_burnt` + gas reserved for later execution.
I don't see a reason why this model cannot be used here. The `used_gas` of the delegate action result after is has been unwrapped includes the usual exec fees of all created actions plus the gas burnt.
2. For function calls, where the exact amount of gas is unknown at the time of gas purchasing, we attach gas. This is treated separately because the user defines this number and it is not fixed in the protocol. But in the `ActionResult`, the attached gas is still included in `used_gas`.
For delegate actions, you could define attached gas to be determined by the inner actions, instead of an explicit field.

Unless there is a good reason for it, is is my opinion that we should avoid introducing the new concept of attached deposit that is used to buy gas a block later.

Instead, I suggest to prepay gas. Just include the gas as prepaid gas here.
https://github.com/near/nearcore/blob/36e3162b0900a8024da2337fe9cd511ad1e7c914/core/primitives/src/transaction.rs#L76-L81

The value for a delegate action should be the sum of all send+exec fees for inner actions. This should balance out `used_gas` for the new receipts, such that `check_balance` does no panic.